src/java.base/share/classes/java/lang/AbstractStringBuilder.java

```diff
@@ -1702,145 +1852,206 @@ final byte getCoder() {
             }
         } else {
-            StringUTF16.putCharsSB(this.value, index, s, off, end);
-        }
-    }
-
-    private void inflateIfNeededFor(String input) {
-        if (COMPACT_STRINGS && (coder != input.coder())) {
-            inflate();
+            StringUTF16.putCharsSB(value, index, s, off, end);
         }
+        return value;
     }
 
-    private void inflateIfNeededFor(AbstractStringBuilder input) {
-        if (COMPACT_STRINGS && (coder != input.getCoder())) {
-            inflate();
+    private static byte[] inflateIfNeededFor(byte[] value, int count, byte coder, byte otherCoder) {
+        if (COMPACT_STRINGS && (coder == LATIN1 && otherCoder == UTF16)) {
+            return inflateToUTF16(value, count);
         }
+        return value;
     }
 
-    private void putStringAt(int index, String str, int off, int end) {
-        inflateIfNeededFor(str);
-        str.getBytes(value, off, index, coder, end - off);
-    }
-
-    private void putStringAt(int index, String str) {
-        inflateIfNeededFor(str);
-        str.getBytes(value, index, coder);
+    /**
+     * {@return the buffer with the substring inserted}
+     * If the substring contains UTF16 characters and the current coder is LATIN1, inflation occurs
+     * into a new buffer and returned; the caller must update the coder to UTF16.
+     * Since the contents are immutable, a simple copy of characters is sufficient.
+     * @param value an existing buffer to insert into
+     * @param coder the coder of the buffer
+     * @param count the count of characters in the buffer
+     * @param index the index to insert the string
+     * @param str the string
+     */
+     private static byte[] putStringAt(byte[] value, byte coder, int count, int index, String str, int off, int end) {
+        byte[] newValue = inflateIfNeededFor(value, count, coder, str.coder());
+        coder = (newValue == value) ? coder : UTF16;
+        str.getBytes(newValue, off, index, coder, end - off);
+        return newValue;
     }
 
-    private final void appendChars(char[] s, int off, int end) {
-        int count = this.count;
-        if (isLatin1()) {
-            byte[] val = this.value;
+    /**
+     * {@return buffer with new characters appended, possibly inflated}
+     * The buffer may need to be inflated if any character does not fit;
+     * if the returned buffer is different then passed in, the new coder is UTF16.
+     * The caller is responsible for updating the count.
+     * @param value the current buffer
+     * @param coder the coder of the buffer
+     * @param count the character count
+     * @param s a char array
+     * @param off the offset of the first character to append
+     * @param end end last (exclusive) character to append
+     */
+    private static byte[] appendChars(byte[] value, byte coder, int count, char[] s, int off, int end) {
+        if (isLatin1(coder)) {
             for (int i = off, j = count; i < end; i++) {
```

Comment:
```suggestion
            int compressed = StringUTF16.compress(s, off, value, count, end - off);
            for (int i = compressed + off, j = count + compressed; i < end; i++) {
```
Refer to PR #24773, adding StringUTF16.compress preprocessing here will improve the performance of java.io.BufferedReader::readLine method.

Should it be done at once or separately for PR #24773 to make changes?

Reply:
I rolled this suggestion into ff3d841; and fixed a second occurence.